### ISSUE TITLE
Consecutive `b` will result in different outcomes when using `in` code generation in specific conditions.

### ISSUE THREAD

**Issue by Klaim** [build2/build2#32]
(I was not super inspired for the issue name, please feel free to change it if you find better)

How to reproduce
--------------------

### A. Quick way:
 1. `git clone https://github.com/Klaim/build2-in-weird.git` (this is a repro case, you can see the setup steps in the history)
 2. `bdep init ...` (I didn't try without this step)
 3. `b` => **leads to error report**: `bug.hxx` missing include refered from `somecpp.cxx`
     - also **generate that `bug.hxx` after the errors** (this is important)
 4. `b` (without cleaning) => **leads to successful build** (because `bug.hxx` is found).

### B. Long way (step by step reproduction):
 1. `bdep new inbug2 -t lib && cd inbug`.
 2. Add `using in` in `build/root.build`.
 3. Add `bug.hxx.in` in `inbug/`. with the content:
```
#pragma once
#include <iostream>
````
4. Add `#include <inbug2/bug.hxx>` in `inbug2/inbug2.cxx`.
5. Add `hxx{bug}: bug/in{bug}` in `inbug2/buildfile`
6. Add a source file in `inbug/somedir/somecpp.cxx` with content (this is the source file that will trigger the issue, under specific conditions):
```
#include "bug.hxx" // The directory for heads is missing on purpose to reproduce the issue.
```
7.  Add the following line in `inbug2/buildfile`:
`cxx.poptions += "-I$out_base" # This is to allow including bug.hxx which is generated in out_base`
8. Do steps A.2. to A.4. to see the error and then success.

Observed
-----------

Linux (Ubuntu):
```
╭─klaim@KlaimLab ~/tests/inbug2 ‹master›
╰─$ b
mkdir ../build-inbug2/inbug2/fsdir{inbug2/}
mkdir ../build-inbug2/inbug2/inbug2/fsdir{somedir/}
error: header 'bug.hxx' not found and cannot be generated
  info: while extracting header dependencies from inbug2/somedir/cxx{somecpp}@../build-inbug2/inbug2/inbug2/somedir/
  info: while applying rule cxx.compile to update ../build-inbug2/inbug2/inbug2/somedir/obja{somecpp}
  info: while applying rule cxx.link to update ../build-inbug2/inbug2/inbug2/liba{inbug2}
  info: while applying rule bin.lib to update ../build-inbug2/inbug2/inbug2/lib{inbug2}
  info: while applying rule alias to update ../build-inbug2/inbug2/dir{inbug2/}
  info: while applying rule alias to update ../build-inbug2/dir{inbug2/}
error: header 'bug.hxx' not found and cannot be generated
  info: while extracting header dependencies from inbug2/somedir/cxx{somecpp}@../build-inbug2/inbug2/inbug2/somedir/
  info: while applying rule cxx.compile to update ../build-inbug2/inbug2/inbug2/somedir/objs{somecpp}
  info: while applying rule cxx.link to update ../build-inbug2/inbug2/inbug2/libs{inbug2}
  info: while applying rule bin.lib to update ../build-inbug2/inbug2/inbug2/lib{inbug2}
  info: while applying rule alias to update ../build-inbug2/inbug2/dir{inbug2/}
  info: while applying rule alias to update ../build-inbug2/dir{inbug2/}
in inbug2/bug/bug.hxx.in
info: failed to update ../build-inbug2/dir{inbug2/}
╭─klaim@KlaimLab ~/tests/inbug2 ‹master›
╰─$ b                                                                                                            1 ↵
mkdir ../build-inbug2/inbug2/tests/fsdir{basics/}
version.in inbug2/version.hxx.in
c++ inbug2/cxx{inbug2}@../build-inbug2/inbug2/inbug2/
c++ inbug2/cxx{inbug2}@../build-inbug2/inbug2/inbug2/
c++ inbug2/somedir/cxx{somecpp}@../build-inbug2/inbug2/inbug2/somedir/
c++ inbug2/somedir/cxx{somecpp}@../build-inbug2/inbug2/inbug2/somedir/
c++ tests/basics/cxx{driver}@../build-inbug2/inbug2/tests/basics/
ar ../build-inbug2/inbug2/inbug2/liba{inbug2}
ld ../build-inbug2/inbug2/inbug2/libs{inbug2}
ld ../build-inbug2/inbug2/tests/basics/exe{driver}
ln ../build-inbug2/inbug2/tests/basics/exe{driver} -> tests/basics/
```

I observe the exact same behavior on Windows. Both versions are the same:
```
╭─klaim@KlaimLab ~/tests/inbug2 ‹master›
╰─$ b --version
build2 0.11.0-a.0.3e8422fe4f99
libbutl 0.11.0-a.0.f3f09dac499f
host x86_64-linux-gnu
Copyright (c) 2014-2019 Code Synthesis Ltd
This is free software released under the MIT license.
```

Expected
----------

Always build? or always fail?

I assume this is a bug because two sequential `b` (without changing any sources) should never result in different outcomes (except no-op for the second if the first succeeded). 
That being said, I think there might be some other incorrectness with this specific setup that I'm not aware of that trigger this issue, feel free to point it to me.

Important Notes
------------------

1. It is obvious from the log that the second `b` succeeds because the first one generated the missing file. What's not normal is that it was not generated before processing the other files dependencies, and it works in other (more common) conditions as you will see below. Also, a side effect is that if we explicitly `b clean` between each `b`, it appears as if the project just can't compile - which can be confusing when there is a massive amount of cpp files including the generated header and the report is therefore very very big and hard to parse (which is normal, not a build2 problem). So it's not always obvious in real world situation.
2. If we add the header directory in `inbug2/somedir/somecpp.cxx` (step B.6.) by replacing it's code by the following code, **there is never any compilation error**:
` #include "inbug2/bug.hxx"`
3. If we move `inbug2/somedir/somecpp.cxx` in `inbug2/`, **there is never any compilation error**. See this log: https://pastebin.com/1kk29c3Y
4. The error report, when there is one, never comes from `inbug2/inbug2.cxx`.
5. This issue is only visible if we use `in`.

(notes 2 and 3 are why it took me so much time to reproduce it)

(I hit this issue while ~hacking~ making a build2 package for libzmq which have one generated header. That's why it's contrived.)

**Comment by boris-kolpackov** [comment]
Thanks for the report. The issue should be fixed in the latest staged toolchain. If you could give it a try and confirm, that would be great.

**Comment by Klaim** [comment]
I checked on windows and ubuntu and it seems to be fixed in stage.

```
PS E:\tools\build2> b --version
build2 0.11.0-a.0.952c3b413992
libbutl 0.11.0-a.0.7fde0c7f2b0c
host x86_64-microsoft-win32-msvc14.2
Copyright (c) 2014-2019 Code Synthesis Ltd
This is free software released under the MIT license.
```
(noting that because the fix commit doesn't refer to this issue)
